File: CSHARP/GraphSearch.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace MtplSample
{
    public class GraphSearch
    {
        private readonly Dictionary<int, List<int>> _adjacency = new();

        public void AddEdge(int from, int to)
        {
            if (!_adjacency.ContainsKey(from)) _adjacency[from] = new List<int>();
            if (!_adjacency.ContainsKey(to)) _adjacency[to] = new List<int>();
            _adjacency[from].Add(to);
            _adjacency[to].Add(from);
        }

        public List<int> Bfs(int start)
        {
            var visited = new HashSet<int>();
            var queue = new Queue<int>();
            var result = new List<int>();

            visited.Add(start);
            queue.Enqueue(start);

            while (queue.Count > 0)
            {
                var current = queue.Dequeue();
                result.Add(current);

                if (!_adjacency.ContainsKey(current)) continue;
                foreach (var neighbor in _adjacency[current])
                {
                    if (visited.Add(neighbor))
                        queue.Enqueue(neighbor);
                }
            }

            return result;
        }

        public List<int> Dfs(int start)
        {
            var visited = new HashSet<int>();
            var result = new List<int>();
            DfsRecursive(start, visited, result);
            return result;
        }

        private void DfsRecursive(int node, HashSet<int> visited, List<int> result)
        {
            visited.Add(node);
            result.Add(node);

            if (!_adjacency.ContainsKey(node)) return;
            foreach (var neighbor in _adjacency[node])
            {
                if (!visited.Contains(neighbor))
                    DfsRecursive(neighbor, visited, result);
            }
        }

        public bool HasPath(int from, int to)
        {
            var visited = new HashSet<int>();
            var stack = new Stack<int>();
            stack.Push(from);

            while (stack.Count > 0)
            {
                var current = stack.Pop();
                if (current == to) return true;
                if (!visited.Add(current)) continue;
                if (!_adjacency.ContainsKey(current)) continue;
                foreach (var neighbor in _adjacency[current])
                {
                    if (!visited.Contains(neighbor))
                        stack.Push(neighbor);
                }
            }
            return false;
        }

        public Dictionary<int, int> ShortestPaths(int start)
        {
            var distances = new Dictionary<int, int>();
            var visited = new HashSet<int>();
            var queue = new Queue<(int Node, int Dist)>();

            queue.Enqueue((start, 0));
            visited.Add(start);

            while (queue.Count > 0)
            {
                var (node, dist) = queue.Dequeue();
                distances[node] = dist;

                if (!_adjacency.ContainsKey(node)) continue;
                foreach (var neighbor in _adjacency[node])
                {
                    if (visited.Add(neighbor))
                        queue.Enqueue((neighbor, dist + 1));
                }
            }

            return distances;
        }
    }
}

File: JS/graphSearch.js
class Graph {
  constructor() {
    this.adjList = new Map();
  }

  addVertex(vertex) {
    if (!this.adjList.has(vertex)) {
      this.adjList.set(vertex, []);
    }
    return this;
  }

  addEdge(v1, v2, weight = 1) {
    if (!this.adjList.has(v1)) this.addVertex(v1);
    if (!this.adjList.has(v2)) this.addVertex(v2);
    this.adjList.get(v1).push({ node: v2, weight });
    this.adjList.get(v2).push({ node: v1, weight });
    return this;
  }

  bfs(start) {
    const visited = new Set();
    const queue = [start];
    const result = [];
    visited.add(start);

    while (queue.length) {
      const vertex = queue.shift();
      result.push(vertex);
      for (const neighbor of this.adjList.get(vertex) || []) {
        if (!visited.has(neighbor.node)) {
          visited.add(neighbor.node);
          queue.push(neighbor.node);
        }
      }
    }
    return result;
  }

  dfs(start) {
    const visited = new Set();
    const result = [];

    const traverse = (vertex) => {
      if (!vertex || visited.has(vertex)) return;
      visited.add(vertex);
      result.push(vertex);
      for (const neighbor of this.adjList.get(vertex) || []) {
        traverse(neighbor.node);
      }
    };

    traverse(start);
    return result;
  }

  hasPath(v1, v2) {
    if (v1 === v2) return true;
    const visited = new Set();
    const queue = [v1];
    visited.add(v1);

    while (queue.length) {
      const vertex = queue.shift();
      for (const neighbor of this.adjList.get(vertex) || []) {
        if (neighbor.node === v2) return true;
        if (!visited.has(neighbor.node)) {
          visited.add(neighbor.node);
          queue.push(neighbor.node);
        }
      }
    }
    return false;
  }

  shortestPath(v1, v2) {
    if (v1 === v2) return [v1];
    const visited = new Set([v1]);
    const queue = [[v1]];

    while (queue.length) {
      const path = queue.shift();
      const vertex = path[path.length - 1];
      for (const neighbor of this.adjList.get(vertex) || []) {
        if (!visited.has(neighbor.node)) {
          const newPath = [...path, neighbor.node];
          if (neighbor.node === v2) return newPath;
          visited.add(neighbor.node);
          queue.push(newPath);
        }
      }
    }
    return null;
  }
}

module.exports = { Graph };
